test/CDFDocumentTest.php

```diff
@@ -0,0 +1,248 @@
+            ->setMethods(['getUuid'])
+            ->getMock();
+        $cdfObjectMockSecond = $this->getMockBuilder(CDFObject::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getUuid'])
+            ->getMock();
+
+        $cdfObjectMockFirst->expects($this->any())
+            ->method('getUuid')
+            ->will($this->returnValue('33333333-0000-0000-0000-000000000000'));
+
+        $cdfObjectMockSecond->expects($this->any())
+            ->method('getUuid')
+            ->will($this->returnValue('44444444-0000-0000-0000-000000000000'));
+
+        return [
+            array_merge($this->providerEntityOperations(), [
+                [
+                    $cdfObjectMockFirst,
+                    $cdfObjectMockSecond
+                ]
+            ])
+        ];
+    }
+
+    public function providerToString()
+    {
+        $cdfObjectMockFirst = $this->getMockBuilder(CDFObject::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getUuid', 'toArray'])
+            ->getMock();
+        $cdfObjectMockSecond = $this->getMockBuilder(CDFObject::class)
+            ->disableOriginalConstructor()
+            ->setMethods(['getUuid', 'toArray'])
+            ->getMock();
+
+        $cdfObjectMockFirst->expects($this->any())
+            ->method('getUuid')
+            ->will($this->returnValue('55555555-0000-0000-0000-000000000000'));
+
+        $cdfObjectMockFirstToArray = [
+            'uuid' => '55555555-0000-0000-0000-000000000000',
+            'type' => 'product',
+            'created' => '2014-12-21T20:12:11+00:00Z',
+            'modified' => '2014-12-21T20:12:11+00:00Z',
+            'origin' => '11111111-0000-0000-0000-000000000000',
+        ];
+
+        $cdfObjectMockSecondToArray = [
+            'uuid' => '66666666-0000-0000-0000-000000000000',
+            'type' => 'product',
+            'created' => '2014-12-21T20:12:11+00:00Z',
+            'modified' => '2014-12-21T20:12:11+00:00Z',
+            'origin' => '22222222-0000-0000-0000-000000000000',
+        ];
+
+        $cdfObjectMockFirst->expects($this->any())
+            ->method('toArray')
+            ->will($this->returnValue($cdfObjectMockFirstToArray));
+
+        $cdfObjectMockSecond->expects($this->any())
+            ->method('getUuid')
+            ->will($this->returnValue('66666666-0000-0000-0000-000000000000'));
+
+        $cdfObjectMockSecond->expects($this->any())
+            ->method('toArray')
+            ->will($this->returnValue($cdfObjectMockSecondToArray));
+
+        return [
+            [
+                [
+                    $cdfObjectMockFirst,
+                    $cdfObjectMockSecond
+                ],
+                json_encode(['entities' => []]),
+                json_encode(['entities' => [
+                    $cdfObjectMockFirstToArray,
+                    $cdfObjectMockSecondToArray
+                ]]),
+            ]
+        ];
+    }
+
+}
```

Comment:
Not sure what you mean by lots of repetitions. Where exactly?
willReturn is a better choice than what?

Reply:
- As for repetitions, for example there are a few instances of `2014-12-21T20:12:11+00:00Z` or `11111111-0000-0000-0000-000000000000` which, in my opinion, could be centralized.

- Instead of using:
``` php
will($this->returnValue('66666666-0000-0000-0000-000000000000'));
```
You could use:
``` php
willReturn('11111111-0000-0000-0000-000000000000');
```

Reply:
- Yes, it could be moved into a single basic array which will be customized afterwards. Which I think is a good idea. But after all these discussions regarding complicated tests I was not sure whether I should do it like that :) But I will.
- As far as I understand it's the same. But ok, willReturn uses less letters, so I will switch to it.

Reply:
No, they are definitely not the same. `will` has more complex stubbing behavior than `willReturn`, which makes it slower.

Reply:
I meant in a context which I'm using it. But what you suggest is a bit better, I agree.